cvmfs/cache_posix.cc

```diff
@@ -293,38 +301,80 @@ void PosixCacheManager::CtrlTxn(
 
 
 string PosixCacheManager::Describe() {
-  return "Posix cache manager (cache directory: " + cache_path_ + ")\n";
+  string msg;
+  if (do_refcount_) {
+    msg = "Refcounting Posix cache manager"
+          "(cache directory: " + cache_path_ + ")\n";
+  } else {
+    msg = "Posix cache manager (cache directory: " + cache_path_ + ")\n";
+  }
+  return msg;
 }
 
 
 /**
- * Nothing to do, the kernel keeps the state of open file descriptors.  Return
- * a dummy memory location.
+ * If not refcounting, nothing to do, the kernel keeps the state 
+ * of open file descriptors.  Return a dummy memory location.
  */
 void *PosixCacheManager::DoSaveState() {
+  if (do_refcount_) {
+    SavedState *state = new SavedState();
+    state->fd_mgr = fd_mgr_->Clone();
+    return state;
+  }
   char *c = reinterpret_cast<char *>(smalloc(1));
-  *c = '\0';
+  *c = kMagicNoRefcount;
   return c;
 }
 
 
 int PosixCacheManager::DoRestoreState(void *data) {
   assert(data);
+  if (do_refcount_) {
+    SavedState *state = reinterpret_cast<SavedState *>(data);
+    if (state->magic_number == kMagicRefcount) {
+      LogCvmfs(kLogCache, kLogDebug, "Restoring refcount cache manager from "
+                                    "refcounted posix cache manager");
+
+      fd_mgr_->AssignFrom(state->fd_mgr.weak_ref());
+    } else {
+      LogCvmfs(kLogCache, kLogDebug, "Restoring refcount cache manager from "
+                                    "non-refcounted posix cache manager");
+    }
+    return -1;
+  }
+
   char *c = reinterpret_cast<char *>(data);
-  assert(*c == '\0');
+  assert(*c == kMagicNoRefcount || *c == kMagicRefcount);
+  if (*c == 123) {
+    SavedState *state = reinterpret_cast<SavedState *>(data);
+    LogCvmfs(kLogCache, kLogDebug, "Restoring non-refcount cache manager from "
+                                    "refcounted posix cache manager - this "
+                                    " is not possible, keep refcounting.");
+    fd_mgr_->AssignFrom(state->fd_mgr.weak_ref());
+    do_refcount_ = true;
+  }
   return -1;
 }
 
 
 bool PosixCacheManager::DoFreeState(void *data) {
+  assert(data);
+  SavedState *state = reinterpret_cast<SavedState *>(data);
+  if (state->magic_number == kMagicRefcount) {
+    delete state;
+  } else {
+  // If not refcounted, the state is the dummy SavedState
+  // of the regular posix cache manager
   free(data);
+  }
```

Comment:
I think here something got accidentally reverted
```suggestion
  }
  // If not refcounted, the state is the dummy SavedState
  // of the regular posix cache manager
  free(data);
```

Reply:
Ah, no, I haven't made this clear, but this is intentional, the reload does not work otherwise. I think it's because only the non-refcounted path in `DoSaveState` calls `smalloc`.

Reply:
Oh, of course! One version calls `new`, the other one `malloc`. Very well then. Then it just requires fixing the indentation in the else block.